Improvements to Hand Constraint juicyness

provencher opened the pull request:
# Overview

Addition of useHybridHandRotation for HandConstraintPalmUp.
- This adds an additional check in HandConstraintPalmUp's hand validation function, which changes the Rotation behavior if useHybridHandRotation is enabled.

Added better behavior for LookAtTrackedObject.
- This introduces a branch in CalculateProjectedSafeZoneRay of the HandConstraint base class, which computes the SafeZone Direction ray according to the rotation of the SolverHandler.TargetTransform, rather than exclusively from the head gaze.

The end result can be seen [here](https://twitter.com/prvncher/status/1211795961510584327)

With useHybridHandRotation enabled:
The menu matches your hand rotation until your palm and head gaze reach a threshold angle, which activates a different rotation behavior that makes the menu look at the camera instead.

In other words, it flips open and closed 
![gUBsKIae9V](https://user-images.githubusercontent.com/7420990/71605193-198aa880-2b35-11ea-92be-3d208075ffca.gif)

keveleigh commented on Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraintPalmUp.cs:
```diff
@@ -55,6 +55,14 @@ public float FlatHandThreshold
 
+        [SerializeField]
+        [Tooltip("When using hybrid hand rotation, solver hand rotation until the menu is aligned with the gaze sufficiently")]
+        private bool useHybridHandRotation = false;
```
Do you think this would make sense as an enum, in case there are other styles like this in the future (or any you can already think of)?
Or does it make more sense as something in addition to other settings?

SimonDarksideJ replied:
An enum might help, but really that would need a system review of this constraint/solver, to ensure all required patterns are handled.  It would mean the implementation would need to change to ensure it is extensible by default.

provencher replied:
Thanks for chiming in @SimonDarksideJ! HandConstraints in general are a bit tacked on right now, as there is no reason for them to be Hand exclusive. I think a general system review would be merited.

Beyond that, there are other conditions I can think of that would affect the rotation mode, or even the overall Solver behavior.

For example, I've written some code that involves keeping a hand menu facing the user, while it's FaceCamera vector is lined up with the GazeVector. 
In other words, even if the palm is no longer facing the camera, the menu will keep facing the camera as long as the user is gazing it. This has the benefit of keep menus stable, even when the menu is fairly large.

Implementing a feature like that would require some rethinking about how hands here.

julenka commented on Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraintPalmUp.cs:
```diff
@@ -55,6 +55,14 @@ public float FlatHandThreshold
+
+        [SerializeField]
+        [Tooltip("Angle between hand up and camera forward, below which the hand menu follows the gaze, if useHybridHandRotation is active.")]
+        private float hybridHandRotationThresholdAngle = 60f;
```
Thanks for these new behaviors! Please also [update the documentation](https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/Contributing/PullRequests.html#documentation-should-be-added-in-the-same-pull-request-as-a-fix--feature) to describe the new fields. The Solvers Page, section [Hand Menu with HandConstraint PalmUp](https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/README_Solver.html#hand-menu-with-handconstraint-and-handconstraintpalmup) would be a great place to describe the new behavior. 

Updating documentation along with the new behaviors helps ensure that the documentation for MRTK stays up to date.

provencher replied:
I just updated the docs!

julenka commented on Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraintPalmUp.cs:
```diff
@@ -94,13 +105,31 @@ protected override bool IsValidController(IMixedRealityController controller)
                     }
+
+                    // Check if the palm angle meets the prescribed threshold
+                    palmFacingThresholdMet = palmCameraAngle < facingThreshold;
```
Interesting, I just had the thought that this could be made more general since currently the logic is "if the palm faces the camera, follow the camera, otherwise follow the tracked object". I think it could be more general by saying "if the tracked object faces the camera, follow the camera, otherwise follow the tracked object". Then it could work for things like motion controllers, other joints, or even following other game objects.

Not required for this PR, but was just interesting.

provencher replied:
I agree that it could work for more than just hands!

Really it's a matter of checking the SolverHandler.TargetTransform - that said, facing the camera for a palm means it's up direction aligns with the gaze vector. 
For a general purpose Solver, it's forward might be more useful. It's difficult to generalize that.

provencher commented:
@julenka Thanks for your review! Let me know if I've addressed all your blocking concerns for this PR.

julenka commented on Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraint.cs:
```diff
@@ -416,8 +416,15 @@ private static Ray CalculateProjectedSafeZoneRay(Vector3 origin, IMixedRealityCo
                     {
-                        direction = Vector3.Cross(CameraCache.Main.transform.forward, Vector3.up);
-                        direction = IsPalmFacingCamera(hand) ? direction : -direction;
+                        if (rotationBehavior == SolverRotationBehavior.LookAtTrackedObject)
```
Could you explain why we are setting direction to transform.right if the behavior is LookAtTrackedObject? Is it a fix for another issue you found?

provencher replied:
Certainly!

So the the reason I changed this is because the follow tracked object rotation behavior sets the position at a constant offset relative to the camera to object vector. 

In my opinion, LookAtTrackedObject should be positioned relative the object's rotation, regardless of the camera's position. This is the meat of this PR to be honest.

However, I realize this may in fact change the behavior MRTK user have come to expect, and so perhaps we should make this change as part of a new rotation behavior enum.

julenka commented on Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraintPalmUp.cs:
```diff
@@ -15,17 +16,18 @@ public class HandConstraintPalmUp : HandConstraint
         /// The angle (in degrees) of the cone between the palm's up and camera's forward have to match. Only supported by <see cref="Microsoft.MixedReality.Toolkit.Input.IMixedRealityHand"/> controllers.
         /// </summary>
-        public float FacingThreshold
+        public float FacingCameraTrackingThreshold
```
I believe this is actually a [breaking change](https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/Contributing/BreakingChanges.html) since it is a change to a public property in a sdk asset. The new naming does seem more descriptive which is a positive. A drawback is that people may get compile errors if they update. This might be a case [where the breaking change adds value but can be written in a way that isn't breaking](https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/Contributing/BreakingChanges.html#the-breaking-change-adds-value-but-could-be-written-in-a-way-that-isnt-breaking).

What are your thoughts on the value vs drawbacks of this rename?

provencher replied:
My mistake! I changed the private value and add the formerly serialized attribute to persist user's previous values in existing prefabs.

To be honest, I don't many folks would be changing this value at runtime via a script. Nothing in the MRTK does so, so it would be entirely in user app level code. If you want me to revert it, I'll do so, but I doubt many folks will notice.

Troy-Ferrell replied:
I think the serialized field can be modified but for the public property make it twice

```
public float FacingCameraTrackingThreshold
{
}
[System.Obsolete("Use FacingCameraTrackingThreshold property instead")
public float FacingThreshold
{
}

julenka commented on Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraintPalmUp.cs:
```diff
@@ -55,6 +57,14 @@ public float FlatHandThreshold
 
+        [SerializeField]
+        [Tooltip("With this active, solver will follow hand rotation until the menu is sufficiently aligned with the gaze, at which point it faces the camera.")]
+        private bool followHandUntilFacingCamera = false;
```
It would be good to make these two fields configurable at runtime, [following the principle 'support configuring in editor and at runtime](https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/Contributing/CodingGuidelines.html#support-configuring-components-both-in-editor-and-at-run-time)

provencher replied:
That makes sense! I can expose them as public properties.

julenka commented on Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraintPalmUp.cs:
```diff
@@ -55,6 +57,14 @@ public float FlatHandThreshold
+
+        [SerializeField]
+        [Tooltip("Angle (in degrees) between hand up and camera forward, below which the hand menu follows the gaze, if followHandUntilFacingCamera is active.")]
+        private float followHandCameraFacingThresholdAngle = 60f;
```
It would be [good to write a test](https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/Contributing/CodingGuidelines.html#prioritize-writing-tests) to ensure this behavior doesn't regress from somebody else's future changes. [Here is some information for how to write an run tests](https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/Contributing/UnitTests.html).

provencher replied:
Hmm... I don't really know how to write a test that might make sense in this case.
Do you think writing a test is a blocking part of this pr?

Troy-Ferrell replied:
I think a basic test would be test starting below the threshold, crossing it, and then crossing back below the threshold. And between these changes validate that the solver positions the object at the expected different places (in this case hand vs camera following)

provencher replied:
If someone writes a test for the hand constraint class, I'll happily add a function to test my changes. I just don't have the time to write up an entire test class for it right now.

provencher commented:
@julenka Thanks for your second review pass! I added the both of you as collaborators.

Troy-Ferrell commented on Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraintPalmUp.cs:
```diff
@@ -15,17 +16,18 @@ public class HandConstraintPalmUp : HandConstraint
-            get { return facingThreshold; }
-            set { facingThreshold = value; }
+            get { return facingCameraTrackingThreshold; }
+            set { facingCameraTrackingThreshold = value; }
```
nit: while here can you switch to using =>
i.e
```
get => facingCameraTrackingThreshold;
set => facingCameraTrackingThreshold = value;
```

provencher replied:
I can but none of the other props in the class use this.

Troy-Ferrell replied:
sorry I meant for all of the props

Troy-Ferrell commented on Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraintPalmUp.cs:
```diff
@@ -94,13 +107,31 @@ protected override bool IsValidController(IMixedRealityController controller)
+                    palmFacingThresholdMet = palmCameraAngle < facingCameraTrackingThreshold;
+
+                    // If using hybrid hand rotation, we proceed with additional checks
+                    if (followHandUntilFacingCamera && palmFacingThresholdMet)
```
nit: Reminder once you create the public properties, use those here and not the private instances

provencher replied:
Done.

Troy-Ferrell commented on Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraint.cs:
```diff
@@ -416,8 +416,15 @@ private static Ray CalculateProjectedSafeZoneRay(Vector3 origin, IMixedRealityCo
+                        {
+                            direction = targetTransform.right;
+                        }
+                        else
```
i feel like here and below should be 
`else if (rotationBehavior == SolverRotationBehavior.LookAtMainCamera)`
?
not sure yet what result should be when SolverRotationBehavior.None?

provencher replied:
@Troy-Ferrell I added another enum here called SolverOffsetBehavior for clarity. There is no none option here so this is no longer an issue. I believe the additional enum adds both clarity and options for users.

provencher commented:
Thanks for the review pass @Troy-Ferrell, I'll try and make some of these additional changes soon. Hopefully some time this week.

provencher commented:
@Troy-Ferrell I updated the code based on the feedback. Can try working on a test.

provencher commented:
Ok @Troy-Ferrell I added the test, and changed the props as you asked.
That said, When I run any playmode test, after the test runs, I get  bunch of errors.

My test seems to run as expected however, so if these playmode tests work for you guys, this new one should behave as expected.

![image](https://user-images.githubusercontent.com/7420990/72559986-12250800-3874-11ea-8562-e624ce3364d8.png)

Troy-Ferrell commented:
> Ok @Troy-Ferrell I added the test, and changed the props as you asked.
> That said, When I run any playmode test, after the test runs, I get bunch of errors.
> 
> My test seems to run as expected however, so if these playmode tests work for you guys, this new one should behave as expected.
> 
> ![image](https://user-images.githubusercontent.com/7420990/72559986-12250800-3874-11ea-8562-e624ce3364d8.png)

That is odd. Nothing looks out-of-ordinary with the test. If it passes in PR validation, then it's fine. 

Things look good. Only last key piece missing is the obsolete property
```c#
[System.Obsolete("Use FacingCameraTrackingThreshold property instead")
public float FacingThreshold
{
...
}
```

Add that in and we can kick off PR validation

provencher commented:
Cheers! Added the prop as requested.

keveleigh commented:
/azp run

provencher commented:
@keveleigh it seems that tests unrelated to my PR have failed. Would re-running solve this?
Also what are the next steps to get this merged in?

keveleigh commented:
@provencher Just triggered a re-run, and filed the failing test at https://github.com/microsoft/MixedRealityToolkit-Unity/issues/7160 so we can make it more consistent.

I'll take a look today about merging this in!
